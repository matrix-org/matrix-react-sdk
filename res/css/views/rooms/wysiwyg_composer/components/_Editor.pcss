/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.mx_WysiwygComposer_Editor_container {
    @keyframes visualbell {
        from {
            background-color: $visual-bell-bg-color;
        }
        to {
            background-color: $background;
        }
    }

    .mx_WysiwygComposer_Editor_content {
        line-height: $font-22px;
        white-space: pre-wrap;
        word-wrap: break-word;
        outline: none;
        overflow-x: hidden;

        /* Force caret nodes to be selected in full so that they can be */
        /* navigated through in a single keypress */
        .caretNode {
            user-select: all;
        }

        // we always have a <br/> tag at the end of the html, we need it to be present at first then hide it as soon as
        // we have any other elements
        br:not(:only-child) {
            display: none;
        }

        p {
            margin-top: 0;
            margin-bottom: 0;
            // this may seem redundant, but we need to handle zero content formatting tags, which occur when we split a
            // formatting tag into paragraphs
            min-height: $font-22px;
        }

        ul,
        ol {
            margin-top: 0;
            margin-bottom: 0;
            padding-inline-start: $spacing-28;
        }

        /* Make list type disc to match rich text editor */
        ul {
            list-style-type: disc;
        }

        blockquote {
            color: #777;
            border-left: 2px solid $blockquote-bar-color;
            border-radius: 2px;
            padding: 0 10px;

            margin-block-start: 0;
            margin-block-end: 0;
            margin-inline-start: 0;
            margin-inline-end: 0;
        }

        > pre {
            font-size: $font-15px;
            line-height: $font-24px;

            margin-top: 0;
            margin-bottom: 0;
            padding: $spacing-8 $spacing-12;

            background-color: $inlinecode-background-color;
            border: 1px solid $inlinecode-border-color;
            border-radius: 2px;
        }

        code {
            font-family: $monospace-font-family !important;
            background-color: $inlinecode-background-color;
            border: 1px solid $inlinecode-border-color;
            border-radius: 4px;
            padding: $spacing-2;
        }

        code:empty {
            border: unset;
            padding: unset;
        }
    }

    .mx_WysiwygComposer_Editor_content_placeholder::before {
        content: var(--placeholder);
        width: 0;
        height: 0;
        overflow: visible;
        display: inline-block;
        pointer-events: none;
        white-space: nowrap;
        color: $tertiary-content;
    }
}
