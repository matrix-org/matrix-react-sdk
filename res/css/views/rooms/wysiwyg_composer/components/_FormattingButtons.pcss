/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.mx_FormattingButtons {
    display: flex;
    justify-content: flex-start;
    gap: 8px;

    .mx_FormattingButtons_Button {
        --size: 28px;
        position: relative;
        cursor: pointer;
        height: var(--size);
        line-height: var(--size);
        width: auto;
        padding-left: 22px;
        background-color: transparent;
        border: none;

        &::before {
            content: '';
            position: absolute;
            top: 6px;
            left: 6px;
            height: 16px;
            width: 16px;
            background-color: $tertiary-content;
            mask-repeat: no-repeat;
            mask-size: contain;
            mask-position: center;
        }

        &::after {
            content: '';
            position: absolute;
            left: 0;
            top: 0;
            z-index: 0;
            width: var(--size);
            height: var(--size);
            border-radius: 5px;
        }

        &:hover {
            &::after {
                background: rgba($secondary-content, 0.1);
            }

            &::before {
                background-color: $secondary-content;
            }
        }
    }

    .mx_FormattingButtons_active {
        &::after {
            background: rgba($accent, 0.1);
        }

        &::before {
            background-color: $accent;
        }
    }

    .mx_FormattingButtons_Button_bold::before {
        mask-image: url('$(res)/img/element-icons/room/composer/bold.svg');
    }

    .mx_FormattingButtons_Button_italic::before {
        mask-image: url('$(res)/img/element-icons/room/composer/italic.svg');
    }

    .mx_FormattingButtons_Button_underline::before {
        mask-image: url('$(res)/img/element-icons/room/composer/underline.svg');
    }

    .mx_FormattingButtons_Button_strikethrough::before {
        mask-image: url('$(res)/img/element-icons/room/composer/strikethrough.svg');
    }
}

.mx_FormattingButtons_Tooltip {
    padding: 0 2px 0 2px;

    .mx_FormattingButtons_Tooltip_KeyboardShortcut {
        color: $tertiary-content;

        kbd {
            margin-top: 2px;
            text-align: center;
            display: inline-block;
            text-transform: capitalize;
            font-size: 12px;
            font-family: Inter, sans-serif;
        }
    }
}
