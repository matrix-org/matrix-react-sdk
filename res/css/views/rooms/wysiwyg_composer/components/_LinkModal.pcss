/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.mx_LinkModal {
    padding: $spacing-32;
    max-width: 600px;
    height: 341px;
    box-sizing: border-box;
    display: flex;
    flex-direction: column;

    .mx_LinkModal_content {
        display: flex;
        flex-direction: column;
        flex: 1;
        gap: $spacing-8;
        margin-top: 7px;

        .mx_LinkModal_Field {
            flex: initial;
            height: 40px;
        }

        .mx_LinkModal_buttons {
            display: flex;
            flex: 1;
            align-items: flex-end;

            .mx_Dialog_buttons {
                display: inline-block;
            }
        }
    }
}
