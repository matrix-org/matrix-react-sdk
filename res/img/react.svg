<svg xmlns="http://www.w3.org/2000/svg" width="18" height="17" viewBox="0 0 18 17">
    <g fill="none" fill-rule="evenodd">
        <path fill="#2E2F32" fill-rule="nonzero" d="M13.6 14.5a7.875 7.875 0 1 1 2.183-3h-.818a7.125 7.125 0 1 0-2.62 3H13.6z"/>
        <g stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width=".75">
            <path d="M11.5 13h6M14.5 10v6"/>
        </g>
        <path stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width=".75" d="M5.5 10s1.125 1.5 3 1.5 3-1.5 3-1.5"/>
        <path stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width="1.5" d="M6.25 6.25h.008M10.75 6.25h.008"/>
    </g>
</svg>
