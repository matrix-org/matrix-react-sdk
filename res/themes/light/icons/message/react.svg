<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><path d="M248 8C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zm0 464c-119.1 0-216-96.9-216-216S128.9 40 248 40s216 96.9 216 216-96.9 216-216 216zm90.2-146.2C315.8 352.6 282.9 368 248 368s-67.8-15.4-90.2-42.2c-5.7-6.8-15.8-7.7-22.5-2-6.8 5.7-7.7 15.7-2 22.5C161.7 380.4 203.6 400 248 400s86.3-19.6 114.8-53.8c5.7-6.8 4.8-16.9-2-22.5-6.8-5.6-16.9-4.7-22.6 2.1zM168 240c17.7 0 32-14.3 32-32s-14.3-32-32-32-32 14.3-32 32 14.3 32 32 32zm160 0c17.7 0 32-14.3 32-32s-14.3-32-32-32-32 14.3-32 32 14.3 32 32 32z"/></svg>