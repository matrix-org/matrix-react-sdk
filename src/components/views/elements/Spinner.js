/*
Copyright 2015, 2016 OpenMarket Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

'use strict';

const React = require('react');

module.exports = React.createClass({
    displayName: 'Spinner',

    render: function() {
        const w = this.props.w || 32;
        const h = this.props.h || 32;
        const imgClass = this.props.imgClassName || "";
        return (
            <div className="mx_Spinner">
                <img src="img/spinner.gif" width={w} height={h} className={imgClass} />
            </div>
        );
    },
});
