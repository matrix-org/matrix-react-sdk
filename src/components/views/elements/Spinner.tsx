/*
Copyright 2015-2021 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from "react";

import { _t } from "../../../languageHandler";

interface IProps {
    w?: number;
    h?: number;
    message?: string;
    onFinished: any; // XXX: Spinner pretends to be a dialog so it must accept an onFinished, but it never calls it
}

export default class Spinner extends React.PureComponent<IProps> {
    public static defaultProps: Partial<IProps> = {
        w: 32,
        h: 32,
    };

    public render(): React.ReactNode {
        const { w, h, message } = this.props;
        return (
            <div className="mx_Spinner">
                {message && (
                    <React.Fragment>
                        <div className="mx_Spinner_Msg">{message}</div>&nbsp;
                    </React.Fragment>
                )}
                <div
                    className="mx_Spinner_icon"
                    style={{ width: w, height: h }}
                    aria-label={_t("common|loading")}
                    role="progressbar"
                    data-testid="spinner"
                />
            </div>
        );
    }
}
