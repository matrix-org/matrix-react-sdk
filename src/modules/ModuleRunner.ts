/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { safeSet } from "matrix-js-sdk/src/utils";
import { TranslationStringsObject } from "@matrix-org/react-sdk-module-api/lib/types/translations";
import { AnyLifecycle } from "@matrix-org/react-sdk-module-api/lib/lifecycles/types";

import { AppModule } from "./AppModule";
import { ModuleFactory } from "./ModuleFactory";

import "./ModuleComponents";

/**
 * Handles and coordinates the operation of modules.
 */
export class ModuleRunner {
    public static readonly instance = new ModuleRunner();

    private modules: AppModule[] = [];

    private constructor() {
        // we only want one instance
    }

    /**
     * Resets the runner, clearing all known modules.
     *
     * Intended for test usage only.
     */
    public reset(): void {
        this.modules = [];
    }

    /**
     * All custom translations from all registered modules.
     */
    public get allTranslations(): TranslationStringsObject {
        const merged: TranslationStringsObject = {};

        for (const module of this.modules) {
            const i18n = module.api.translations;
            if (!i18n) continue;

            for (const [lang, strings] of Object.entries(i18n)) {
                safeSet(merged, lang, merged[lang] || {});

                for (const [str, val] of Object.entries(strings)) {
                    safeSet(merged[lang], str, val);
                }
            }
        }

        return merged;
    }

    /**
     * Registers a factory which creates a module for later loading. The factory
     * will be called immediately.
     * @param factory The module factory.
     */
    public registerModule(factory: ModuleFactory): void {
        this.modules.push(new AppModule(factory));
    }

    /**
     * Invokes a lifecycle event, notifying registered modules.
     * @param lifecycleEvent The lifecycle event.
     * @param args The arguments for the lifecycle event.
     */
    public invoke(lifecycleEvent: AnyLifecycle, ...args: any[]): void {
        for (const module of this.modules) {
            module.module.emit(lifecycleEvent, ...args);
        }
    }
}
