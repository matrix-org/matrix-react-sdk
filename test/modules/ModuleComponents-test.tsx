/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from "react";
import { render } from "@testing-library/react";
import { TextInputField } from "@matrix-org/react-sdk-module-api/lib/components/TextInputField";
import { Spinner as ModuleSpinner } from "@matrix-org/react-sdk-module-api/lib/components/Spinner";

import "../../src/modules/ModuleRunner";

describe("Module Components", () => {
    // Note: we're not testing to see if there's components that are missing a renderFactory()
    // but rather that the renderFactory() for components we do know about is actually defined
    // and working.
    //
    // We do this by deliberately not importing the ModuleComponents file itself, relying on the
    // ModuleRunner import to do its job (as per documentation in ModuleComponents).

    it("should override the factory for a TextInputField", () => {
        const { asFragment } = render(<TextInputField label="My Label" value="My Value" onChange={() => {}} />);
        expect(asFragment()).toMatchSnapshot();
    });

    it("should override the factory for a ModuleSpinner", () => {
        const { asFragment } = render(<ModuleSpinner />);
        expect(asFragment()).toMatchSnapshot();
    });
});
