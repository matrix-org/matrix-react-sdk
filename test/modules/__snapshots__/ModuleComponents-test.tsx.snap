// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Module Components should override the factory for a ModuleSpinner 1`] = `
<Spinner>
  <Spinner
    h={32}
    w={32}
  >
    <div
      className="mx_Spinner"
    >
      <div
        aria-label="Loading..."
        className="mx_Spinner_icon"
        role="progressbar"
        style={
          Object {
            "height": 32,
            "width": 32,
          }
        }
      />
    </div>
  </Spinner>
</Spinner>
`;

exports[`Module Components should override the factory for a TextInputField 1`] = `
<TextInputField
  label="My Label"
  onChange={[Function]}
  value="My Value"
>
  <Field
    autoComplete="off"
    element="input"
    label="My Label"
    onChange={[Function]}
    type="text"
    validateOnBlur={true}
    validateOnChange={true}
    validateOnFocus={true}
    value="My Value"
  >
    <div
      className="mx_Field mx_Field_input"
    >
      <input
        autoComplete="off"
        id="mx_Field_1"
        label="My Label"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="My Label"
        type="text"
        value="My Value"
      />
      <label
        htmlFor="mx_Field_1"
      >
        My Label
      </label>
    </div>
  </Field>
</TextInputField>
`;
